user-operator/src/main/java/io/strimzi/operator/user/operator/batching/AbstractBatchReconciler.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.user.operator.batching;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+/**
+ *
+ */
+public abstract class AbstractBatchReconciler<T> implements Runnable {
+    private final static Logger LOGGER = LogManager.getLogger(AbstractBatchReconciler.class);
+
+    private final BlockingQueue<T> queue;
+    private final int maxBatchSize;
+    private final int maxBatchTime;
+    private final Thread controllerThread;
+
+    private volatile CountDownLatch microBatch;
+    private volatile boolean stop = false;
+
+    public AbstractBatchReconciler(String name, int queueSize, int maxBatchSize, int maxBatchTime) {
+        if (maxBatchSize > queueSize)   {
+            throw new RuntimeException("Maximum batch size cannot be bigger than queue size");
+        }
+
+        this.queue = new ArrayBlockingQueue<>(queueSize);
+        this.microBatch = new CountDownLatch(0);
+        this.maxBatchSize = maxBatchSize;
+        this.maxBatchTime = maxBatchTime;
+        this.controllerThread = new Thread(this, name);
+    }
+
+    protected abstract void reconcile(Collection<T> items);
+
+    public void enqueue(T item) throws InterruptedException {
+        queue.put(item);
+
+        if (queue.size() >= maxBatchSize)   {
+            microBatch.countDown();
+        }
+    }
+
+    @Override
+    public void run() {
+        LOGGER.info("{}: BatchReconciler is running", controllerThread.getName());
+
+        while (!stop)    {
+            try {
+                LOGGER.trace("{}: Waiting for the next batch of the BatchReconciler", controllerThread.getName());
```

Comment:
I'll mention it here, but it applies all over: It would be easier (and better, I think) just to configure the default logging pattern to include the thread name, rather passing it explicitly as an argument everywhere.

Reply:
TBH, I do not like logging with thread name in the pattern. I prefer this way ... and I do not think this is used in too many places here. There are many log messages not doing this where it doesn't matter.